tests/test_container_provider.py

```diff
@@ -0,0 +1,121 @@
+import pytest
+from mock import Mock
+
+from ansible.module_utils.basic import AnsibleModule
+
+from miqclient.api import API
+import manageiq
+
+
+@pytest.fixture(autouse=True)
+def miq_api_class(monkeypatch):
+    miq_api_class = Mock(spec=API)
+    monkeypatch.setattr("manageiq.MiqApi", miq_api_class)
+    yield miq_api_class
+
+
+@pytest.fixture
+def the_provider():
+    the_provider = Mock()
+    the_provider.name = "The provider"
+    yield the_provider
+
+
+@pytest.fixture
+def miq_ansible_module():
+    miq_ansible_module = Mock(spec=AnsibleModule)
+    yield miq_ansible_module
+
+
+@pytest.fixture
+def endpoints(miq):
+    endpoints = miq.generate_endpoints("The hostname", "12345", "The token",
+                                       "The HHostname", "54321")
+    yield endpoints
+
+
+class AnsibleModuleFailed(Exception):
+    pass
+
+
+@pytest.fixture()
+def miq(miq_api_class, miq_ansible_module, the_provider):
+    miq_ansible_module.params = {'metrics': True}
+
+    def fail(msg):
+        raise AnsibleModuleFailed(msg)
+
+    miq_ansible_module.fail_json = fail
+    miq = manageiq.ManageIQ(miq_ansible_module, "http://themanageiq.tld",
+                            "The username", "The password")
+
+    miq_api_class.return_value.post.return_value = dict(results=[dict(id=0)])
+    miq_api_class.return_value.get.return_value = dict(endpoints=[
+        {'port': '12345', 'role': 'default', 'hostname': 'The hostname'}])
+    miq_api_class.return_value.collections.providers = [the_provider]
+
+    yield miq
+
+
+def test_generate_endpoints_with_metrics(miq):
+    endpoints = miq.generate_endpoints("The hostname", "12345", "The token",
+                                       "The HHostname", "54321")
+    assert endpoints == [
+        {'authentication': {'auth_key': 'The token', 'role': 'bearer'},
+         'endpoint': {'hostname': 'The hostname',
+                      'port': '12345',
+                      'role': 'default'}},
+        {'authentication': {'auth_key': 'The token', 'role': 'hawkular'},
+         'endpoint': {'hostname': 'The HHostname',
+                      'port': '54321',
+                      'role': 'hawkular'}}
+    ]
+
+
+def test_generate_endpoints_without_metrics(miq):
+    miq.module.params = {'metrics': False}
+    endpoints = miq.generate_endpoints("The hostname", "12345", "The token",
+                                       object(), object())
```

Comment:
another case where tests reveal the code being suboptimal :-)
I think in actual usage these 2 params are None when unspecified in playbook.